KAFKA-10847: Add a RocksDBTimeOrderedWindowStore to hold records using their timestamp as key prefix

spena opened the pull request:
The new `RocksDBTimeOrderedWindowStore` stores key/value records with a combined key that starts with the record timestamp, followed by a sequential number and the key in bytes (`<time,seq,key>`). This key combination will keep all records ordered by timestamp in the RocksDB store. The advantage of keeping records this way is to make range queries functions more efficient, like `fetchAll(from, to)`. This range method returns all keys in the specified time range using a more efficient key schema to iterate on RocksDB.

`RocksDBTimeOrderedWindowStore` with key query ranges, like `fetch(key, from, to)` or `fetch(keyFrom, keyTo, from, to)`, do not perform very well. For cases with key query ranges, it is better to use the current `RocksDBWindowStore` which stores records using the record key as the RocksDB key prefix. `RocksDBTimeOrderedWindowStore` is meant to fix the issue with https://issues.apache.org/jira/browse/KAFKA-10847 which requires a temporary store where to hold non-joined records, and later do a query range of all keys in a specified time range.

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

- Added unit tests

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

spena commented:
Failing test is `kafka.server.ListOffsetsRequestTest.testResponseIncludesLeaderEpoch()` which it is unrelated to this PR.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/Stores.java:
```diff
@@ -37,6 +37,7 @@
 
 import static org.apache.kafka.streams.internals.ApiUtils.prepareMillisCheckFailMsgPrefix;
 import static org.apache.kafka.streams.internals.ApiUtils.validateMillisecondDuration;
+import static org.apache.kafka.streams.state.internals.RocksDbWindowBytesStoreSupplier.WindowStoreTypes;
```
Since `Stores` are public APIs, we would need to file a KIP in order to change it. On the other hand, `Stores` is used by users to customize their materialized state stores, while for KAFKA-10847 we can just hard-code which types of stores to use not through `Stores` factory.

spena replied:
Right. I'll remove this.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/TimeOrderedWindowStoreBuilder.java:
```diff
@@ -0,0 +1,80 @@
+
+import java.util.Objects;
+
+public class TimeOrderedWindowStoreBuilder<K, V> extends AbstractStoreBuilder<K, V, WindowStore<K, V>> {
```
If we do not allow such stores to be created from `Stores`, maybe we could remove this class as well.

spena replied:
I'm not sure about this one yet. I found this issue in the `RocksDBTimeOrderedWindowStoreTest` which tests the store can be restored from a changelog. The only different thing with this builder is the `maybeWrapLogging` that returns an instance of `ChangeLoggingTimeOrderedWindowBytesStore`.

I believe we need to allow the store to be restored, right? If I used the generic `ChangeLoggingWindowBytesStore`, then the restore fails because the generic uses the `WindowKeySchema`.

spena replied:
I just removed the `ChangeLoggingTimeOrderedWindowBytesStore` class in this new PR, and pass the window key serializer to the `ChangeLoggingWindowBytesStore` to reuse it. But I think I still need the `TimeOrderedWindowStoreBuilder`.

guozhangwang replied:
Makes sense. I realized this store cannot be removed now.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentIterator.java:
```diff
@@ -82,9 +93,9 @@ public boolean hasNext() {
                 } else {
                     if (forward) {
-                        currentIterator = currentSegment.range(from, to);
+                        currentIterator = currentSegment.range(from, to, prefixScan);
```
Instead of passing this boolean around and creating overloaded interfaces, could we:

* add `prefixScan` in Segment / SegmentIterator.
* add a different `RocksDBPrefixIterator` besides `RocksDBRangeIterator` which would be used in `prefixScan`, and then the `RocksDBStore#prefixScan` to be used in SegmentIterator.

guozhangwang replied:
I should clarify: not meant the existing `RocksDBStore#prefixScan` since it only allows a single prefix, but adding another function.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/Segment.java:
```diff
@@ -25,4 +26,35 @@
+     * @return The reverse iterator for this range, from largest to smallest key bytes.
+     * @throws NullPointerException       If null is used for from or to.
+     */
+    KeyValueIterator<Bytes, byte[]> reverseRange(Bytes from, Bytes to, boolean prefixScan);
```
Related to the other comment: since in stream-stream join we do not really need reverse-prefixScan, just adding a forward `prefixScan(..)` interface may be better.

guozhangwang reviewed (commented):
Hi Sergio, thanks for the PR! I have some meta comments below:

1. Let's consider breaking it up to smaller ones for ease of reviews. I'm thinking maybe we can just have:
1.a. new TimeOrderedStateStore impl (along with corresponding unit tests, same below).
1.b. new prefixScan iterator with new with bytes comparator.
1.c. (not yet in this PR) improved `KStreamsImplJoin` that add the additional time ordered state store.

2. The extended `AbstractWindowBytesStoreTest` only have a single `shouldFetchAllInTimeRange` that does not cover some edge cases of `fetchAll` calls we may have in stream-stream join. We should consider improving its test coverage.

guozhangwang commented:
cc @cadonna to take a look around prefix scan, and @vcrfxia for the bytes layout.

spena commented:
Thanks @guozhangwang for the feedback. I will create new PRs.

spena committed:
KAFKA-10847: Add new RocksDBTimeOrderedWindowStore that persists (time-key)-value records

This new store is more efficient when calling range queries with only time parameters, like
`fetch(from, to)`. For range queries using key ranges, then the current RocksDBWindowStore should
be used.

spena commented:
@guozhangwang I just replace this PR with only the changes for the `RocksDBTimeOrderedWindowStore`. I will create other PRs with your suggestions.

guozhangwang commented:
> @guozhangwang I just replace this PR with only the changes for the `RocksDBTimeOrderedWindowStore`. I will create other PRs with your suggestions.

I will continue reviewing this PR, as I understand you would not `replace` it with other PRs.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/WindowKeySchemaTest.java:
```diff
@@ -242,7 +242,7 @@ public void shouldConvertToBinaryAndBack() {
 
     @Test
-    public void shouldExtractEndTimeFromBinary() {
+    public void shouldExtractSequenceFromBinary() {
```
Thank you! :)

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/OrderedBytes.java:
```diff
@@ -65,4 +65,16 @@ static Bytes lowerRange(final Bytes key, final byte[] minSuffix) {
         );
     }
+
+    static Bytes range(final byte[] prefix, final Bytes key) {
```
nit: rename to `prefixRange`?

guozhangwang replied:
Please let me know if you want to keep it as is -- which is fine.

But I think we should try to avoid byte array copies as much as possible: right now we first do a byte array for the prefix, in the schema, and then allocate a longer byte array copying the prefix, and then the key bytes. We can, instead, moving this logic directly to the schema where we just create a single array once, and put in the timestamp / sequence / key correspondingly

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSegmentedBytesStore.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    private RocksDBTimeOrderedSegmentedBytesStore(final String name,
+                                          final String metricsScope,
```
nit: indentations.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractRocksDBSegmentedBytesStoreTest.java:
```diff
@@ -93,7 +93,7 @@
     @Parameters(name = "{0}")
     public static Object[] getKeySchemas() {
-        return new Object[] {new SessionKeySchema(), new WindowKeySchema()};
+        return new Object[] {new SessionKeySchema(), new WindowKeySchema(), new TimeOrderedKeySchema()};
```
Should `TimeOrderedKeySchema` only be used for `RocksDBTimeOrderedSegmentedBytesStoreTest`? In this way the added schema would be used for other extended test classes as well.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSegmentedBytesStore.java:
```diff
@@ -0,0 +1,57 @@
+    private final KeySchema keySchema;
+    private final AbstractSegments<KeyValueSegment> segments;
+
+    RocksDBTimeOrderedSegmentedBytesStore(final String name,
```
I think this store would only work with `TimeOrderedKeySchema` and no others, right? If yes, then we can just create the schema internally at construction, and no need to ask for a pass-in parameter.

spena replied:
I'll remove this class. I added because I was going to overwrite the `fetchAll` to use prefixes to fetch data. Now that we don't need to do that, then this class is useless. I will re-use the `RocksDBSegmentedBytesStore` and pass the `TimeOrderedKeySchema` to it.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSegmentedBytesStore.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @Override
+    public KeyValueIterator<Bytes, byte[]> fetchAll(final long timeFrom,
```
Why need to override here? Seems exactly the same as `AbstractRocksDBSegmentedBytesStore`.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSegmentedBytesStoreTest.java:
```diff
@@ -0,0 +1,36 @@
+    private final static String METRICS_SCOPE = "metrics-scope";
+
+    RocksDBTimeOrderedSegmentedBytesStore getBytesStore() {
+        return new RocksDBTimeOrderedSegmentedBytesStore(
```
This test would cover for all three types of `schema` in the `AbstractRocksDBSegmentedBytesStoreTest`, is this intentional? See my other comment below.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedWindowStore.java:
```diff
@@ -0,0 +1,145 @@
+/**
+ * A persistent (time-key)-value store based on RocksDB.
+ */
+public class RocksDBTimeOrderedWindowStore
```
Could we declare this class as `extends RocksDBWindowStore` so that we can save on those functions like `init` as their logics do not change, also no duplicated `maybeUpdateSeqnumForDups`.

guozhangwang replied:
Actually as I think twice I think we can just implement `put` and `all`, and for other read operations directly throw unsupported exceptions to make sure if they are called by mistake, we can fail fast. Also by doing that, we can just make `TimeOrderedWindowStoreIteratorWrapper` an inner class since it is only used by `all()`

spena replied:
Seems too much work on the testing side if I keep most of the methods unsupported. Btw, I will remove the `RocksDBTimeOrderedSegmentedBytesStore` because we don't need to do range queries with prefixes anymore. With that, all methods in this class will work as expected without the issues with the range iterators caused by looking with prefixed keys. I think we should keep all the functionality. It is harmless, and it is already tested by `RocksDBTimeOrderedWindowStoreTest`. What do you think?

guozhangwang replied:
Yeah that's a good point about unit test complexities. Let's just emphasize that key ranges are sub-optimally implemented with the new schema then (see my other comment below).

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/TimeOrderedKeySchema.java:
```diff
@@ -0,0 +1,211 @@
+    private static final int SEQNUM_SIZE = 4;
+    private static final int PREFIX_SIZE = TIMESTAMP_SIZE + SEQNUM_SIZE;
+
+    public static Bytes lowerTimeRange(final long from) {
```
Following my previous comments: we can be more strict to not support any other operations than `put` and `all`, and as a result we can also just not support these `XXrange` APIs as well but only `toStoreKeyBinary / extractStoreTimestamp` for puts and get deserialization. Basically, we want to be eliminate vulnerability to bugs from those not expected function calls

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractWindowBytesStoreTest.java:
```diff
@@ -1173,14 +1173,23 @@ private void putSecondBatch(final WindowStore<Integer, String> store,
         store.put(2, "two+6");
     }
 
+    long extractStoreTimestamp(final byte[] binaryKey) {
```
About test coverage: for `all()` function today we only have a single `shouldNotThrowConcurrentModificationException`. Let's add some scenarios as well:

1) deleted records (create some that would be at the head of the all iterators, mimicking the stream-stream join behavior) would not be returned from `all`.
2) returned iterators can be closed early, and subsequent `all` would still return the whole set.
3) Ordering of `all` are indeed lexico for prefix -- that's the key we would leverage on for stream-stream join.

spena replied:
Added tests cases

guozhangwang reviewed (commented):
Made a second pass. So far it looks very promising!

spena committed:
Address PR feedback

spena committed:
Add more all() unit tests

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/TimeOrderedKeySchema.java:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    @Override
+    public Bytes upperRangeFixedSize(final Bytes key, final long to) {
```
We know these functions are triggered by `fetch(final Bytes key, final long timeFrom, final long timeTo)` and following the default implementation it is sub-optimal since we will range over a large scan and then drop a lot of the records. Let's add a oneliner comment on top of them referring readers to the head javadoc of this schema class that they should try avoid ever calling these functions.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/TimeOrderedKeySchema.java:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    @Override
+    public HasNextCondition hasNextCondition(final Bytes binaryKeyFrom, final Bytes binaryKeyTo, final long from, final long to) {
```
As discussed before, for `fetchAll(final long timeFrom, final long timeTo)` we actually do not need to trigger this function at all since we know it should always return true.

I think we can either 1) claim that `fetchAll(final long timeFrom, final long timeTo)` is also not optimal and people should avoid using it with the new schema, or 2) try to still keep that impl as optimal as possible, i.e. in `AbstractRocksDBSegmentedBytesStore#fetchAll` we have a condition like this:

```
return keySchema instanceOf TimeOrderedKeySchema ?
            return new SegmentIterator<>(
            searchSpace.iterator(),
            (....) -> true,
            TimeOrderedKeySchema.toStoreKeyBinary(0, from, 0),
            TimeOrderedKeySchema.toStoreKeyBinary(0, to + 1, Integer.MAX_VALUE),
            true) : // else return the normal implementation
```

guozhangwang reviewed (commented):
Thanks for the added tests! Just a few more comments and otherwise LGTM!

guozhangwang commented:
@spena could you also take a look at the failed tests and see if they are related to this PR?

spena committed:
Address PR feedback
- Add comments about optimization for TimeOrderedKeySchema

spena commented:
@guozhangwang The test failures are not related to this PR.
```
Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.JoinWithIncompleteMetadataIntegrationTest.testShouldAutoShutdownOnJoinWithIncompleteMetadata
```

The time-ordered store is not used anywhere yet, so this failed test is 100% not caused by this PR>

guozhangwang commented:
> @guozhangwang The test failures are not related to this PR.

Sounds good. I think they are irrelevant indeed, similarly the other failures from `KStreamRepartitionIntegrationTest` and `KTableKTableForeignKeyInnerJoinMultiIntegrationTest`.

guozhangwang commented:
I've triggered another round of unit tests to see if they are transient -- will review the latest commit now.

spena committed:
Fix javadoc comment

guozhangwang commented:
Merged to trunk, thanks @spena !!
